xarray/tests/test_accessor_str.py

```diff
@@ -224,10 +322,893 @@ def test_replace_literal(dtype):
+                "a_fifth_b_5555_c_opqr",
+                "a_sixth_b_66666_c_rst",
+            ],
+        ],
+        dims=["X", "Y"],
+    ).astype(dtype)
+
+    with pytest.raises(KeyError):
+        value.str.extract(pat=pat_str, dim="X")
+
+    with pytest.raises(KeyError):
+        value.str.extract(pat=pat_re, dim="X")
+
+    with pytest.raises(KeyError):
+        value.str.extractall(pat=pat_str, group_dim="X", match_dim="ZZ")
+
+    with pytest.raises(KeyError):
+        value.str.extractall(pat=pat_re, group_dim="X", match_dim="YY")
+
+    with pytest.raises(KeyError):
+        value.str.extractall(pat=pat_str, group_dim="XX", match_dim="Y")
+
+    with pytest.raises(KeyError):
+        value.str.extractall(pat=pat_re, group_dim="XX", match_dim="Y")
+
+    with pytest.raises(KeyError):
+        value.str.extractall(pat=pat_str, group_dim="ZZ", match_dim="ZZ")
+
+    with pytest.raises(KeyError):
+        value.str.extractall(pat=pat_re, group_dim="ZZ", match_dim="ZZ")
+
+
+def test_extract_single_case(dtype):
+    pat_str = r"(\w+)_Xy_\d*"
+    conv = {np.unicode_: str, np.bytes_: lambda x: bytes(x, encoding="UTF-8")}[dtype]
+    pat_re = re.compile(conv(pat_str))
+
+    value = xr.DataArray(
+        [
+            ["a_Xy_0", "ab_xY_10-bab_Xy_110-baab_Xy_1100", "abc_Xy_01-cbc_Xy_2210"],
+            [
+                "abcd_Xy_-dcd_Xy_33210-dccd_Xy_332210",
+                "",
+                "abcdef_Xy_101-fef_Xy_5543210",
+            ],
+        ],
+        dims=["X", "Y"],
+    ).astype(dtype)
+
+    targ_none = xr.DataArray(
+        [["a", "bab", "abc"], ["abcd", "", "abcdef"]], dims=["X", "Y"]
+    ).astype(dtype)
+    targ_dim = xr.DataArray(
+        [[["a"], ["bab"], ["abc"]], [["abcd"], [""], ["abcdef"]]], dims=["X", "Y", "XX"]
+    ).astype(dtype)
+
+    res_str_none = value.str.extract(pat=pat_str, dim=None)
+    res_str_dim = value.str.extract(pat=pat_str, dim="XX")
+    res_str_none_case = value.str.extract(pat=pat_str, dim=None, case=True)
+    res_str_dim_case = value.str.extract(pat=pat_str, dim="XX", case=True)
+    res_re_none = value.str.extract(pat=pat_re, dim=None)
+    res_re_dim = value.str.extract(pat=pat_re, dim="XX")
+
+    assert res_str_none.dtype == targ_none.dtype
+    assert res_str_dim.dtype == targ_dim.dtype
+    assert res_str_none_case.dtype == targ_none.dtype
+    assert res_str_dim_case.dtype == targ_dim.dtype
+    assert res_re_none.dtype == targ_none.dtype
+    assert res_re_dim.dtype == targ_dim.dtype
+
+    assert_equal(res_str_none, targ_none)
+    assert_equal(res_str_dim, targ_dim)
+    assert_equal(res_str_none_case, targ_none)
+    assert_equal(res_str_dim_case, targ_dim)
+    assert_equal(res_re_none, targ_none)
+    assert_equal(res_re_dim, targ_dim)
+
+
+def test_extract_single_nocase(dtype):
+    pat_str = r"(\w+)_Xy_\d*"
+    conv = {np.unicode_: str, np.bytes_: lambda x: bytes(x, encoding="UTF-8")}[dtype]
+    pat_re = re.compile(conv(pat_str), flags=re.IGNORECASE)
```

Comment:
I think this should be easier to understand:
```suggestion
    pat_str = pat_str if dtype is np.str_ else bytes(pat_str, encoding="UTF-8")
    pat_re = re.compile(pat_str, flags=re.IGNORECASE)
```
---
(Alternatively:



```suggestion
    pat_str = {np.unicode_: pat_str, np.bytes_: bytes(pat_str, encoding="UTF-8")}[dtype]
    pat_re = re.compile(pat_str, flags=re.IGNORECASE)
```
)

Reply:
Fixed